File: crates/workspace-core/src/workspace.rs
use std::collections::{HashMap, HashSet};
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};

use crate::files;
use crate::layout::{default_layout, extract_terminal_ids};
use crate::terminal::TerminalSession;

type TabId = u32;

#[derive(Clone, Serialize, Deserialize)]
pub struct TabInfo {
    pub id: TabId,
    pub title: String,
    pub layout_json: String,
    pub root_path: String,
}

#[derive(Clone, Serialize, Deserialize)]
pub struct WorkspaceState {
    pub tabs: Vec<TabInfo>,
    pub active_tab_id: TabId,
}

pub struct Workspace {
    /// Root new tabs are seeded with. Each tab can then repoint its own
    /// `root_path` independently via `set_tab_root_path` — this is only
    /// the starting point for tabs created after that point.
    pub default_root_path: PathBuf,
    tabs: Vec<Tab>,
    active_tab_id: TabId,
    terminals: HashMap<u32, TerminalSession>,
    next_terminal_id: u32,
    next_tab_id: TabId,
}

struct Tab {
    id: TabId,
    title: String,
    layout_json: String,
    root_path: PathBuf,
}

impl Workspace {
    pub fn new() -> Self {
        Self::with_root(std::env::current_dir().unwrap_or_else(|_| PathBuf::from(".")))
    }

    /// `root`'s own validity isn't checked here — the caller (the
    /// persisted-config loader in the Tauri app) is responsible for
    /// falling back to the CWD default if the saved path no longer
    /// exists.
    pub fn with_root(root: PathBuf) -> Self {
        let mut ws = Self {
            default_root_path: root,
            tabs: Vec::new(),
            active_tab_id: 0,
            terminals: HashMap::new(),
            next_terminal_id: 0,
            next_tab_id: 0,
        };
        ws.add_tab();
        ws
    }

    /// Rebuilds a `Workspace` from a previously-persisted `WorkspaceState`
    /// (see `AppConfig`/`workspace.json` in `src/lib.rs`) instead of
    /// starting with a single fresh tab. Reuses every tab's original id
    /// (rather than reassigning fresh ones via `add_tab`) and spawns a
    /// terminal for each id actually referenced in that tab's
    /// `layout_json` — this is what makes `TerminalSession`'s tmux
    /// session-key-by-id reattach to the *same* session it had before
    /// instead of colliding with, or losing track of, another tab's.
    /// Falls back to today's fresh-tab behavior if the snapshot is empty
    /// (e.g. first launch, no `workspace.json` yet).
    pub fn from_snapshot(root: PathBuf, snapshot: WorkspaceState) -> Self {
        if snapshot.tabs.is_empty() {
            return Self::with_root(root);
        }

        let mut next_tab_id = 0;
        let mut next_terminal_id = 0;
        let mut tabs = Vec::with_capacity(snapshot.tabs.len());
        let mut terminals = HashMap::new();

        for tab in &snapshot.tabs {
            next_tab_id = next_tab_id.max(tab.id + 1);
            let tab_root = PathBuf::from(&tab.root_path);
            let tab_root = if tab_root.is_dir() {
                tab_root
            } else {
                root.clone()
            };

            for terminal_id in extract_terminal_ids(&tab.layout_json) {
                next_terminal_id = next_terminal_id.max(terminal_id + 1);
                let session = TerminalSession::new(terminal_id, 120, 40, Some(tab_root.clone()));
                session.start();
                terminals.insert(terminal_id, session);
            }

            tabs.push(Tab {
                id: tab.id,
                title: tab.title.clone(),
                layout_json: tab.layout_json.clone(),
                root_path: tab_root,
            });
        }

        let active_tab_id = if tabs.iter().any(|t| t.id == snapshot.active_tab_id) {
            snapshot.active_tab_id
        } else {
            tabs[0].id
        };

        Self {
            default_root_path: root,
            tabs,
            active_tab_id,
            terminals,
            next_terminal_id,
            next_tab_id,
        }
    }

    pub fn add_tab(&mut self) -> TabId {
        let id = self.next_tab_id;
        self.next_tab_id += 1;
        let root_path = self.default_root_path.clone();
        let terminal_id = self.spawn_terminal_in(&root_path, 120, 40);
        let layout_json = default_layout(terminal_id);
        self.tabs.push(Tab {
            id,
            title: format!("Tab {}", id + 1),
            layout_json,
            root_path,
        });
        self.active_tab_id = id;
        id
    }

    pub fn close_tab(&mut self, tab_id: TabId) -> Result<(), String> {
        if self.tabs.len() <= 1 {
            return Err("cannot close the last tab".into());
        }
        let Some(idx) = self.tabs.iter().position(|t| t.id == tab_id) else {
            return Err("tab not found".into());
        };

        let tab = self.tabs.remove(idx);
        self.release_terminals_only_in_tab(&tab.layout_json);

        if self.active_tab_id == tab_id {
            let next = idx.min(self.tabs.len().saturating_sub(1));
            self.active_tab_id = self.tabs[next].id;
        }

        Ok(())
    }

    pub fn select_tab(&mut self, tab_id: TabId) {
        if self.tabs.iter().any(|t| t.id == tab_id) {
            self.active_tab_id = tab_id;
        }
    }

    pub fn set_tab_layout(&mut self, tab_id: TabId, layout_json: String) {
        let Some(tab) = self.tabs.iter_mut().find(|t| t.id == tab_id) else {
            return;
        };

        let old_ids: HashSet<u32> = extract_terminal_ids(&tab.layout_json).into_iter().collect();
        tab.layout_json = layout_json.clone();
        let new_ids: HashSet<u32> = extract_terminal_ids(&layout_json).into_iter().collect();

        for id in old_ids.difference(&new_ids) {
            if !self.is_terminal_referenced(*id) {
                self.terminals.remove(id);
            }
        }
    }

    pub fn spawn_terminal(&mut self, cols: u16, rows: u16) -> u32 {
        let root = self
            .tabs
            .iter()
            .find(|t| t.id == self.active_tab_id)
            .map(|t| t.root_path.clone())
            .unwrap_or_else(|| self.default_root_path.clone());
        self.spawn_terminal_in(&root, cols, rows)
    }

    fn spawn_terminal_in(&mut self, root: &Path, cols: u16, rows: u16) -> u32 {
        let id = self.next_terminal_id;
        self.next_terminal_id += 1;
        let session = TerminalSession::new(id, cols, rows, Some(root.to_path_buf()));
        session.start();
        self.terminals.insert(id, session);
        id
    }

    /// Existing terminals keep their own cwd — only ones spawned after
    /// this takes effect. Rejects anything that isn't an existing
    /// directory rather than silently falling back, so a typo'd Settings
    /// path doesn't quietly leave you where you were. Scoped to a single
    /// tab — other tabs keep whatever root they already had.
    pub fn set_tab_root_path(&mut self, tab_id: TabId, path: PathBuf) -> Result<(), String> {
        if !path.is_dir() {
            return Err(format!("not a directory: {}", path.display()));
        }
        let Some(tab) = self.tabs.iter_mut().find(|t| t.id == tab_id) else {
            return Err("tab not found".into());
        };
        tab.root_path = path;
        Ok(())
    }

    pub fn tab_root_path(&self, tab_id: TabId) -> Option<PathBuf> {
        self.tabs
            .iter()
            .find(|t| t.id == tab_id)
            .map(|t| t.root_path.clone())
    }

    pub fn terminal_write(&self, id: u32, data: &[u8]) {
        if let Some(session) = self.terminals.get(&id) {
            session.write(data);
        }
    }

    pub fn terminal_resize(&mut self, id: u32, cols: u16, rows: u16) {
        if let Some(session) = self.terminals.get_mut(&id) {
            session.resize(cols, rows);
        }
    }

    pub fn poll_all_terminals(&self) -> Vec<(u32, Vec<u8>)> {
        let mut outputs = Vec::new();
        for (id, session) in &self.terminals {
            let chunks = session.drain_chunks();
            for chunk in chunks {
                if !chunk.is_empty() {
                    outputs.push((*id, chunk));
                }
            }
        }
        outputs
    }

    pub fn state(&self) -> WorkspaceState {
        WorkspaceState {
            tabs: self
                .tabs
                .iter()
                .map(|t| TabInfo {
                    id: t.id,
                    title: t.title.clone(),
                    layout_json: t.layout_json.clone(),
                    root_path: t.root_path.to_string_lossy().into_owned(),
                })
                .collect(),
            active_tab_id: self.active_tab_id,
        }
    }

    pub fn list_dir(&self, tab_id: TabId, rel: &str) -> Result<Vec<files::DirEntry>, String> {
        files::list_dir(&self.tab_root(tab_id)?, rel)
    }

    pub fn read_file(&self, tab_id: TabId, rel: &str) -> Result<String, String> {
        files::read_file(&self.tab_root(tab_id)?, rel)
    }

    pub fn write_file(&self, tab_id: TabId, rel: &str, content: &str) -> Result<(), String> {
        files::write_file(&self.tab_root(tab_id)?, rel, content)
    }

    fn tab_root(&self, tab_id: TabId) -> Result<PathBuf, String> {
        self.tabs
            .iter()
            .find(|t| t.id == tab_id)
            .map(|t| t.root_path.clone())
            .ok_or_else(|| "tab not found".into())
    }

    fn is_terminal_referenced(&self, terminal_id: u32) -> bool {
        self.tabs
            .iter()
            .any(|t| extract_terminal_ids(&t.layout_json).contains(&terminal_id))
    }

    fn release_terminals_only_in_tab(&mut self, layout_json: &str) {
        for id in extract_terminal_ids(layout_json) {
            if !self.is_terminal_referenced(id) {
                self.terminals.remove(&id);
            }
        }
    }
}

File: electron/src/main/workspace.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { Pty } from "./pty";
import { PtySession } from "./ptySession";
import { defaultLayout, extractTerminalIds } from "./layout";
import { salvageLayoutJson } from "../shared/layoutSalvage";
import * as files from "./files";
import type { DirEntry } from "./files";
import * as search from "./search";
import { MEDIA_MIME_TYPES, toMediaUrl } from "./mediaProtocol";
import { openEpub, type EpubBook } from "./epub";
import { toEngineBundleUrl } from "./engineBundlePaths";
import { registerProjectApp as registerProjectApp_ } from "./projectManifest";
import type { ProjectManifest } from "../shared/projectManifest";

// Direct port of crates/workspace-core/src/workspace.rs.

export interface TabInfo {
  id: number;
  title: string;
  layoutJson: string;
  rootPath: string;
}

export interface WorkspaceState {
  tabs: TabInfo[];
  activeTabId: number;
}

interface Tab {
  id: number;
  title: string;
  layoutJson: string;
  rootPath: string;
}

interface TerminalEntry {
  session: PtySession;
}

export class Workspace {
  /** Root new tabs are seeded with. Each tab can then repoint its own
   * rootPath independently via setTabRootPath — this is only the
   * starting point for tabs created after that point. */
  defaultRootPath: string;
  private tabs: Tab[] = [];
  private activeTabId = 0;
  private terminals = new Map<number, TerminalEntry>();
  private nextTerminalId = 0;
  private nextTabId = 0;
  /** Fired whenever a terminal produces output for an attached renderer. */
  onTerminalData: ((id: number, seq: number, data: Buffer) => void) | null = null;

  private constructor(rootPath: string) {
    this.defaultRootPath = rootPath;
  }

  static withRoot(rootPath: string): Workspace {
    const ws = new Workspace(rootPath);
    ws.addTab();
    return ws;
  }

  /** Rebuilds a Workspace from a previously-persisted WorkspaceState
   * instead of starting with a single fresh tab. Reuses every tab's
   * original id and spawns a terminal for each id actually referenced in
   * that tab's layoutJson. PtySession replay restores output while the
   * app is running; across a full app restart, each terminal instead
   * reattaches to its own tmux session (see pty.ts) — the shell and
   * everything running in it survived the restart on the OS side, so
   * there's no scrollback to replay, just a live reattach. */
  static fromSnapshot(rootPath: string, snapshot: WorkspaceState): Workspace {
    if (snapshot.tabs.length === 0) return Workspace.withRoot(rootPath);

    const ws = new Workspace(rootPath);
    let nextTabId = 0;
    let nextTerminalId = 0;

    for (const tab of snapshot.tabs) {
      nextTabId = Math.max(nextTabId, tab.id + 1);
      const tabRoot = isDir(tab.rootPath) ? tab.rootPath : rootPath;
      const { json: layoutJson } = salvageLayoutJson(tab.layoutJson);

      for (const terminalId of extractTerminalIds(layoutJson)) {
        nextTerminalId = Math.max(nextTerminalId, terminalId + 1);
        ws.spawnTerminalWithId(terminalId, 120, 40, tabRoot);
      }

      ws.tabs.push({ id: tab.id, title: tab.title, layoutJson, rootPath: tabRoot });
    }

    ws.nextTabId = nextTabId;
    ws.nextTerminalId = nextTerminalId;
    ws.activeTabId = ws.tabs.some((t) => t.id === snapshot.activeTabId)
      ? snapshot.activeTabId
      : ws.tabs[0].id;

    return ws;
  }

  addTab(): number {
    const id = this.nextTabId++;
    const rootPath = this.defaultRootPath;
    const terminalId = this.spawnTerminalIn(rootPath, 120, 40);
    this.tabs.push({ id, title: `Tab ${id + 1}`, layoutJson: defaultLayout(terminalId), rootPath });
    this.activeTabId = id;
    return id;
  }

  closeTab(tabId: number): void {
    if (this.tabs.length <= 1) throw new Error("cannot close the last tab");
    const idx = this.tabs.findIndex((t) => t.id === tabId);
    if (idx === -1) throw new Error("tab not found");

    const [tab] = this.tabs.splice(idx, 1);
    this.releaseTerminalsOnlyInTab(tab.layoutJson);

    if (this.activeTabId === tabId) {
      const next = Math.min(idx, this.tabs.length - 1);
      this.activeTabId = this.tabs[next].id;
    }
  }

  selectTab(tabId: number): void {
    if (this.tabs.some((t) => t.id === tabId)) this.activeTabId = tabId;
  }

  renameTab(tabId: number, title: string): void {
    const tab = this.tabs.find((t) => t.id === tabId);
    if (!tab) throw new Error("tab not found");
    const trimmed = title.trim();
    if (!trimmed) throw new Error("title cannot be empty");
    tab.title = trimmed;
  }

  /** Reorders tabs to match `orderedIds` — must be exactly the current
   * tab ids, just permuted (a drag-and-drop reorder never adds/removes a
   * tab). Silently ignored if it isn't, rather than partially applying a
   * stale drag payload from before a tab closed mid-drag. */
  reorderTabs(orderedIds: number[]): void {
    if (orderedIds.length !== this.tabs.length) return;
    const byId = new Map(this.tabs.map((t) => [t.id, t]));
    const reordered = orderedIds.map((id) => byId.get(id));
    if (reordered.some((t) => !t)) return;
    this.tabs = reordered as Tab[];
  }

  setTabLayout(tabId: number, layoutJson: string): void {
    const tab = this.tabs.find((t) => t.id === tabId);
    if (!tab) return;

    const oldIds = new Set(extractTerminalIds(tab.layoutJson));
    tab.layoutJson = layoutJson;
    const newIds = new Set(extractTerminalIds(layoutJson));

    for (const id of oldIds) {
      if (!newIds.has(id) && !this.isTerminalReferenced(id)) {
        this.terminals.get(id)?.session.disposeAndDestroySession();
        this.terminals.delete(id);
      }
    }
  }

  spawnTerminal(cols: number, rows: number): number {
    const root = this.tabs.find((t) => t.id === this.activeTabId)?.rootPath ?? this.defaultRootPath;
    return this.spawnTerminalIn(root, cols, rows);
  }

  private spawnTerminalIn(root: string, cols: number, rows: number): number {
    const id = this.nextTerminalId++;
    this.spawnTerminalWithId(id, cols, rows, root);
    return id;
  }

  private spawnTerminalWithId(id: number, cols: number, rows: number, root: string): void {
    const pty = new Pty({ cols, rows, cwd: root });
    pty.start();
    const session = new PtySession(id, pty, cols, rows);
    session.setOnData((terminalId, seq, data) => {
      this.onTerminalData?.(terminalId, seq, data);
    });
    this.terminals.set(id, { session });
  }

  connectTerminal(id: number, webContentsId: number) {
    const entry = this.terminals.get(id);
    if (!entry) throw new Error("terminal not found");
    return entry.session.connect(webContentsId);
  }

  disconnectTerminal(id: number, webContentsId: number): void {
    const entry = this.terminals.get(id);
    if (!entry) return;
    entry.session.disconnect(webContentsId);
  }

  getTerminalSession(id: number): PtySession | undefined {
    return this.terminals.get(id)?.session;
  }

  /** Existing terminals keep their own cwd — only ones spawned after this
   * takes effect. Rejects anything that isn't an existing directory. */
  setTabRootPath(tabId: number, rootPath: string): void {
    if (!isDir(rootPath)) throw new Error(`not a directory: ${rootPath}`);
    const tab = this.tabs.find((t) => t.id === tabId);
    if (!tab) throw new Error("tab not found");
    tab.rootPath = rootPath;
  }

  tabRootPath(tabId: number): string | undefined {
    return this.tabs.find((t) => t.id === tabId)?.rootPath;
  }

  terminalWrite(id: number, data: Buffer): void {
    this.terminals.get(id)?.session.write(data);
  }

  terminalResize(id: number, cols: number, rows: number): void {
    const entry = this.terminals.get(id);
    if (!entry) return;
    entry.session.resize(cols, rows);
  }

  state(): WorkspaceState {
    return {
      tabs: this.tabs.map((t) => ({ id: t.id, title: t.title, layoutJson: t.layoutJson, rootPath: t.rootPath })),
      activeTabId: this.activeTabId,
    };
  }

  listDir(tabId: number, rel: string): DirEntry[] {
    return files.listDir(this.tabRoot(tabId), rel);
  }
  readFile(tabId: number, rel: string): string {
    return files.readFile(this.tabRoot(tabId), rel);
  }
  readFileBinaryPreview(tabId: number, rel: string): files.BinaryFilePreview | null {
    return files.readFileBinaryPreview(this.tabRoot(tabId), rel);
  }
  writeFile(tabId: number, rel: string, content: string): void {
    files.writeFile(this.tabRoot(tabId), rel, content);
  }
  createDir(tabId: number, rel: string): void {
    files.createDir(this.tabRoot(tabId), rel);
  }
  deletePath(tabId: number, rel: string): void {
    files.deletePath(this.tabRoot(tabId), rel);
  }
  renamePath(tabId: number, fromRel: string, toRel: string): void {
    files.renamePath(this.tabRoot(tabId), fromRel, toRel);
  }

  searchInFiles(
    tabId: number,
    query: string,
    opts: search.SearchOptions,
    onFile: (result: search.SearchFileResult) => void,
    onDone: (error?: string) => void,
  ): search.ActiveSearch {
    return search.searchInFiles(this.tabRoot(tabId), query, opts, onFile, onDone);
  }
  listAllFiles(tabId: number): Promise<string[]> {
    return search.listAllFiles(this.tabRoot(tabId));
  }

  /** Every currently-open tab's root — the media protocol handler confines
   * requests against all of them, not just the tab active when playback
   * started (a media file's owning tab may not be the visible one). */
  allTabRootPaths(): string[] {
    return [this.defaultRootPath, ...this.tabs.map((t) => t.rootPath)];
  }

  mediaUrl(tabId: number, rel: string): string | null {
    const ext = path.extname(rel).toLowerCase();
    if (!(ext in MEDIA_MIME_TYPES)) return null;
    const resolved = files.resolveUnderRoot(this.tabRoot(tabId), rel);
    return toMediaUrl(resolved);
  }

  openEpub(tabId: number, rel: string): Promise<EpubBook> {
    return openEpub(this.tabRoot(tabId), rel);
  }

  /** `rel` is a workspace-relative *directory* holding an already-built
   * engine Web export (index.html + its .js/.wasm/.pck siblings) — see
   * engineBundleProtocol.ts. Resolved through the same
   * files.resolveUnderRoot confinement every other file op here uses. */
  engineBundleUrl(tabId: number, rel: string, entry?: string): string {
    const resolved = files.resolveUnderRoot(this.tabRoot(tabId), rel);
    return toEngineBundleUrl(resolved, entry);
  }

  /** Registers one app/document entry into `tabId`'s project (keyed by
   * its rootPath) — see projectManifest.ts. `rel` is stored as-given
   * (workspace-relative), not resolved to an absolute path, so the entry
   * stays valid if the project ever moves on disk. */
  registerProjectApp(tabId: number, kind: string, rel: string, title?: string): ProjectManifest {
    return registerProjectApp_(this.tabRoot(tabId), { id: rel, kind, path: rel, title });
  }

  disposeAllTerminals(): void {
    for (const entry of this.terminals.values()) entry.session.dispose();
    this.terminals.clear();
  }

  private tabRoot(tabId: number): string {
    const root = this.tabs.find((t) => t.id === tabId)?.rootPath;
    if (!root) throw new Error("tab not found");
    return root;
  }

  private isTerminalReferenced(terminalId: number): boolean {
    return this.tabs.some((t) => extractTerminalIds(t.layoutJson).includes(terminalId));
  }

  private releaseTerminalsOnlyInTab(layoutJson: string): void {
    for (const id of extractTerminalIds(layoutJson)) {
      if (!this.isTerminalReferenced(id)) {
        this.terminals.get(id)?.session.disposeAndDestroySession();
        this.terminals.delete(id);
      }
    }
  }
}

function isDir(p: string): boolean {
  try {
    return fs.statSync(p).isDirectory();
  } catch {
    return false;
  }
}
